staker/block_validator.go

```diff
@@ -490,527 +426,436 @@ func (v *BlockValidator) SetCurrentWasmModuleRoot(hash common.Hash) error {
-			if err != nil {
-				log.Warn("failed to get message in block validator", "err", err)
-				return
-			}
-			status, err := v.newValidationStatus(prevHeader, header, msg)
+func (v *BlockValidator) iterativeValidationEntryCreator(ctx context.Context, ignored struct{}) time.Duration {
+	moreWork, err := v.createNextValidationEntry(ctx)
+	if err != nil {
+		processed, processedErr := v.streamer.GetProcessedMessageCount()
+		log.Error("error trying to create validation node", "err", err, "created", v.created()+1, "processed", processed, "processedErr", processedErr)
+	}
+	if moreWork {
+		return 0
+	}
+	return v.config().ValidationPoll
+}
+
+func (v *BlockValidator) sendNextRecordPrepare() error {
+	if v.nextRecordPrepared != nil {
+		if v.nextRecordPrepared.Ready() {
+			prepared, err := v.nextRecordPrepared.Current()
 			if err != nil {
-				log.Warn("failed to create validation status", "err", err)
-				return
+				return err
 			}
-			v.blockMutex.Lock()
-			entry, found = v.validations.Load(nextRecord)
-			if !found {
-				v.validations.Store(nextRecord, status)
-				entry = status
+			if prepared > v.prepared {
+				v.prepared = prepared
 			}
-			v.blockMutex.Unlock()
+			v.nextRecordPrepared = nil
+		} else {
+			return nil
 		}
-		validationStatus, ok := entry.(*validationStatus)
-		if !ok || (validationStatus == nil) {
-			log.Error("bad entry trying to send recordings")
-			return
-		}
-		currentStatus := validationStatus.getStatus()
-		if currentStatus == RecordFailed {
-			// retry
-			v.validations.Delete(nextRecord)
-			v.triggerSendValidations()
-			return
-		}
-		if currentStatus == Unprepared {
-			prevHeader := validationStatus.Entry.PrevBlockHeader
-			if prevHeader != nil {
-				_, err := v.recordingDatabase.GetOrRecreateState(ctx, prevHeader, stateLogFunc)
-				if err != nil {
-					log.Error("error trying to prepare state for recording", "err", err)
-				}
-				// add another reference that will be released by the record thread
-				_, err = v.recordingDatabase.StateFor(prevHeader)
-				if err != nil {
-					log.Error("error trying re-reference state for recording", "err", err)
-				}
-				if v.lastHeaderForPrepareState != nil {
-					v.recordingDatabase.Dereference(v.lastHeaderForPrepareState)
-				}
-				v.lastHeaderForPrepareState = prevHeader
-			}
-			err := v.sendRecord(validationStatus, true)
-			if err != nil {
-				log.Error("error trying to send preimage recording", "err", err)
-			}
-		}
-		nextRecord++
 	}
+	nextPrepared := v.validated() + arbutil.MessageIndex(v.config().PrerecordedBlocks)
+	created := v.created()
+	if nextPrepared > created {
+		nextPrepared = created
+	}
+	if v.prepared >= nextPrepared {
+		return nil
+	}
+	nextPromise := containers.NewPromise[arbutil.MessageIndex](nil)
+	v.LaunchThread(func(ctx context.Context) {
+		err := v.recorder.PrepareForRecord(ctx, v.prepared, nextPrepared-1)
+		if err != nil {
+			nextPromise.ProduceError(err)
+		} else {
+			nextPromise.Produce(nextPrepared)
```

Comment:
Could we call sendRecord() here for each of the blocks we prepared? Especially if this is no longer happening in a background thread.